File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        mp = defaultdict(int)
        left = 0
        mx_len = 0
        
        for right in range(len(s)):
            while s[right] in mp:
                mp[s[left]] -= 1
                if not mp[s[left]]:
                    del mp[s[left]]
                left += 1

            mp[s[right]] += 1
            mx_len = max(mx_len, right-left+1)
        
        return mx_len

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int i=0, j=0, ml=-1;
        HashMap<Character, Integer> mp = new HashMap<>();

        while(j<s.length()) {
            mp.put(s.charAt(j), mp.getOrDefault(s.charAt(j), 0) + 1);

            if(mp.size() == j-i+1) {
                ml = Math.max(ml, j-i+1);
            } else if(mp.size() < j-i+1) {
                while(mp.size() < j-i+1) {
                    mp.put(s.charAt(i), mp.get(s.charAt(i))-1);
                    if(mp.get(s.charAt(i)) == 0) {
                        mp.remove(s.charAt(i));
                    }
                    i++;
                }
            }
            j++;
        }
        if(ml>0) return ml;
        else return 0;
    }
}